### ISSUE TITLE
test(metalgemm): isolate graph Q8 quantization in a non-fast-math library

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#12919]
Parent: #12349
Coordinates with: #12853
Coordinates with: #12914

## Parent context
Original exact Metal MTP parity remains RED. #12914 directly compared all1024 activation codes and32 scale bit patterns on native Apple M3 Pro: codes identical, three scales differ. All four normalized rows are exact; H/D/S projected outputs remain pairwise unequal. H-versus-S under fixed host quantization is a separate route discrepancy.

## Current state
Frozen #12914 candidate: q4k.m SHA25680aac5a9a02ee46fda08b09b78c9e3fd2c1ceb8086af34e791fb418f78f07946, full manifestaa7283af34c9b53a40c08c39ebae4311672c9b956889c38af1fee13b95b63313, strict testd682fe65cde687886a0368f3c6bea8eed2e6b5d03343b3b7f165212010742b13. Physical receipt1a33c28ca1e0cdda0fef5709d11b2d64a10490968d0548f2b6008b928e59c598, executor56092 exit1, one run/zero skips. First scale mismatch row1/token5/block5 host3cd63ec4/device3cd63ec3.

## For / Problem / Today / Better because
For exact native Metal qualification, test whether isolated quantizer compilation semantics explain scale-bit drift. Today graph_quantize_q8 shares the Q4/Q6 library compiled with options:nil. A private library with fastMathEnabled=NO changes one kernel's compile regime while preserving its source algorithm and all other shader options.

## Why this is next
CPU arm64 quant_quantize_arm64.s:64 uses FDIVS amax/127. Graph q4k.m:724 uses a/127.0f but q4k_init compiles its library with defaults. Apple documents default fast math and its permission to change floating-point results. This is a concrete hypothesis, not proof that compiler math mode caused all observed drift.

## Prior art and official documentation (observed 2026-09-12)
Applied sota-check. Captured installed `fak sota list --json` matrix and path/slug human lookup; the installed path/slug command ignored trailing --json, so do not describe that output as machine JSON. Matching row metal-quant-gemm names llama.cpp Metal/MLX and cpuref cosine1.000000. This experiment retains the stronger existing bitwise parent oracle.

- Apple fastMathEnabled API: https://developer.apple.com/documentation/metal/mtlcompileoptions/fastmathenabled?changes=_2_1&language=objc . Defaults to YES; allows optimizations that may violate IEEE754; deprecated in favor of mathMode. Local installed Metal SDK MTLLibrary.h:305-320 independently confirms default and macOS10.11 availability/macOS15 deprecation. No availability migration is bundled here.
- Apple WWDC2016 Advanced Metal Shader Optimization, fast-math section: https://developer.apple.com/videos/play/wwdc2016/606/ . Describes changed precision guarantees, reassociation and contraction. Historical event2016, checked2026-09-12. This supports the experiment, not a universal exact-division guarantee on the present device.
- Matrix primary reference https://github.com/ggml-org/llama.cpp inspected at revision3057bb66c86c46d5781e50e85462a760ba7d1feb. `ggml/src/ggml-metal/CMakeLists.txt`:143-151 uses -fno-fast-math and -fno-inline for GGML_METAL_SHADER_DEBUG, with a backend-test motivation. This is a debug-specific precedent, not a claim of its production default or a direct equivalent. `ggml-metal-device.m` uses per-library MTLCompileOptions. Exact LICENSE is MIT, SHA25694f29bbed6a22c35b992c5c6ebf0e7c92f13b836b90f36f461c9cf2f0f1d010d. No external expressive code is copied; retain FAK kernel bytes unchanged. GitHub content API supplied pinned files after local urllib certificate validation failed; no TLS bypass used.
- Official MSL PDF fetch through web returned an internal error; normative full-spec coverage is unavailable in this pass. No unverified spec guarantee is asserted. Broad upstream issue/PR/release mining is outside this bounded compile-option question.

Route: stay-minimal. Portfolio disposition WATCH for isolated strict quantizer compilation until physical evidence; current shipping default remains unchanged. Global fast-math disable, GEMM replacement and engine binding are EXCLUDE for this one-file experiment. Candidate concerns exact P4/H256/M3Pro parity; added support cost is one runtime library compile and one isolated compile-options object. Review trigger is this exact witness and later SDK changes. Disconfirming result is persistent scale drift or failure to initialize/execute.

## Working spine
Move the existing graph_quantize_q8 kernel and its existing comments verbatim into a new private kGraphQuantizeQ8Src string in q4k.m, with only metal_stdlib/namespace preamble. Close kQ4KSrc before it. Compile kQ4KSrc with existing options:nil; compile only kGraphQuantizeQ8Src using a private MTLCompileOptions with fastMathEnabled=NO. Bind only psoGraphQuantizeQ8 from that library. No kernel statement, reduction, division expression, rounding expression, argument, launch shape, or oracle changes.

## Proposed diff and initialization failure
Unapplied exact proposal `/tmp/fak-quant-precise-proposal/proposed-unapplied.patch`, SHA25604207a3aeb09753cc94a0b10dbf4f4f8fe6ec5b644aa3d632200dd1afdf27ae9. No repository source has been edited.

Allocate dedicated options; if nil log and return0. Compile dedicated library with its own NSError; if nil log and return0. Resolve graph_quantize_q8; if nil log and return0. Create local quantPipeline; if nil log and return0. Assign psoGraphQuantizeQ8 only after success. Existing required-pipeline gate and gQ4KReady assignment remain unchanged. There is no attempt to compile quantization again using default fast math and no new alternate engine.

The quantizer is already a required q4k_init pipeline: failure keeps ready0 and returns initialization failure to existing callers, potentially making Q4/Q6 operations unavailable as well. The proposal preserves that coupling; it does not redesign init/rollback or change existing higher-level unsupported behavior. A skipped or unexecuted diagnostic is not acceptance. Existing graph inputs/buffers/receipts are unchanged.

## Core through-line
One compile-option isolation, unchanged kernel algorithm, existing direct intermediate witness and full strict oracle. The regime change affects every operation in the quantizer, including code division/rounding; attribution to a single compiler transformation would require additional evidence.

## Scope
```routing
lane: metalgemm
paths: ["internal/metalgemm/q4k.m"]
expected_steps: 3
```

## Likely files
internal/metalgemm/q4k.m

## Scope class
Work unit: leaf. One S1 source-only compile candidate; proposal review before edits, canonical admission and managed tree before implementation.

## Definition of done
- [ ] Independent proposal/init-failure/diff/budget review accepts this candidate.
- [ ] Exact one-file source candidate independently reviewed and frozen.
- [ ] Separately granted native witness compares all32 scales/all1024 codes, H/D, H/S and full strict result separately.
- [ ] Independent physical review records scoped disposition without claiming parent parity or sole cause.

## Gold-plating boundary
No broad compiler flags, mathMode migration, shader rewrite, fma/rounding changes, tolerance, test edits, added device runs, library swap or performance promotion. Existing parent and diagnostic source remain unchanged. Proposal only until reviewed.

## Acceptance gate
Same TestMetalMTPP4StrictSerialParity, P4/H256/M3Pro, original Q8 projection route. Require actual native execution, one test and zero skips. Compare all1024 codes and32 scale bits with existing hashes/counts/first-difference output. Check norm rows and H/D first; report H/S separately because host-quantized projection-route drift is independent. Preserve all strict logits/GDN/KV/hidden/continuation assertions and actual overall exit. Diagnostic/accounting remains1152 separate bytes, production20480-read/6272-upload/four encoders. Equality is measured, not assumed. Full raw arrays are not emitted by the frozen diagnostic.

## Done condition
Captured experiment disposition and independently reviewed evidence; original #12349 acceptance remains unchanged and open until all original gates pass.

## Witness
env -u FAK_METAL_MPS CGO_ENABLED=1 GOMAXPROCS=2 GOFLAGS=-p=1 GOCACHE=/tmp/fak-12853-reduction-run/cache GOTMPDIR=/tmp/fak-12853-reduction-run/tmp go test ./internal/model -run '^TestMetalMTPP4StrictSerialParity$' -count=1 -timeout=60s -json

## Budget
No CPU/HIL grant now. After source review, one separately granted attempt: test60, dedicated owned group TERM230/KILL240, cleanup census270 and final receipt/release by300. Source/owner drift cancels the packet; no reprepare or retry inside that grant. Extra runtime shader-library compilation is inside the same budget; prior build/run41.738s is context, not a cost guarantee. No benchmark or throughput claim.

## Duplicate search
Open/closed public queries `Metal quantization fast math` and `"graph_quantize_q8"` found adjacent #9192 and diagnostic #12914; no dedicated quantizer-library math-options candidate.

## Problem frame
Centrality: Core
Portfolio tier: 2 serving
P1: advanced - exact native quantizer diagnosis.
P2: preserved - no unmeasured performance claim.
P3: preserved - one private library in one file.
P4: advanced - original physical oracle unchanged.

## Work estimate
Estimate: 1 point.

## Overall completion contribution
Contribution: 1/4 points within #12349; planning only.

## Completion standard
experiment

## Closure binding
Explicit-path witnessed commit cites child and (fak metalgemm), with Prior-art trailer naming Apple Metal compile options and the pinned llama.cpp debug precedent, or independently record a non-landing experiment disposition. No claim that parent parity is complete.

**Comment by anthony-chaudhary** [comment]
Approved proposal applied exactly in managed worker `fak-worker-wt-metalgemm-fa700762879c`; source-only candidate frozen, no build/test/HIL/commit.

- Applied patch SHA256 `04207a3aeb09753cc94a0b10dbf4f4f8fe6ec5b644aa3d632200dd1afdf27ae9` exactly matches reviewed proposal.
- q4k.m SHA256 `f7c3bdc617c66dda92acf0e5614ff0495502c5c6edff23520396eff3354a54df`.
- Full1702-entry source manifest `36122cb1852de431cdd7e04f943cad8468c5dee4dc7151479426124850307d51`.
- Original #12914 strict diagnostic test remains `d682fe65cde687886a0368f3c6bea8eed2e6b5d03343b3b7f165212010742b13`.
- Programmatic comparison confirms quantizer statements/comments unchanged, other library options:nil unchanged, and exactly one source-path delta relative to frozen #12914. Apply/diff checks pass.
- Artifacts `_scratch/issue-12919-source/{candidate.patch,source.sha256,receipt.json,before-q4k.m,approved-proposal.md}`. Original #12914 physical artifacts preserved. Canonical one-path lease released.

Next: final source-identity review, then a separately granted single physical witness. Whole-quantizer compilation remains a hypothesis; exact code/scale results, H/D, H/S and full strict outcome must be distinguished. No arithmetic or performance success claim.

**Comment by anthony-chaudhary** [comment]
One physical #12919 candidate witness completed: actual executor5765 exit1, native Apple M3 Pro, one test/zero skips. Patch04207a/q4kf7c3/full36122/testd682 unchanged across all1702 source entries. Owned group12273 and recorded descendants absent at terminal; canonical lease released; no retry.

Direct result on this exact P4/H256 fixture:
- All1024 codes match exactly; all32 scale bit patterns now match exactly (previous diagnostic had three scale mismatches). Common code SHA256cdfecda26215f8bef90149c1afb481689f2723facdbbeb7237a0f2f592e7e588; common scale SHA25600f8cb88a3a83b03c56be3fffbd306054c226dc350491462a33381bd081e04fb.
- All four normalized rows remain exact. Token7 layer0 mixed-QKV H and D now match: SHA256f2b0f3b38465ff8a36ce4ede53656d3c242f5860c95225461efc8285c5d19a3e.
- H-versus-S remains different. Serial S remains SHA256cf28fa88ba3dbc73633df3c321ed8127753dfb02f84d31f300c02b0786f7deb0; channel0 H/D bf7e14ef versus S bf7e14f0.
- Lifecycle rejection and owned-copy checks pass. Separate diagnostic1152bytes; production one read20480bytes/upload6272bytes/four encoders, total observed21632readbytes.
- Full strict overall parity remains RED for logits/GDN/KV/hidden/continuation. No oracle or tolerance changed. The whole-quantizer compilation change supports only this fixture; it does not identify a sole compiler transformation or establish general exactness/performance.

Build emitted the expected fastMathEnabled deprecation warning and duplicate-lobjc linker warning; initialization and native execution succeeded. Test1.27s/package1.794s and supervised61.480s are operational records only. Independently observed peer landing PID13018 started12:11:31 PDT, overlapping the candidate window12:10:52–12:11:53 and test12:11:51.546832–12:11:52.813724. Coordinator also reported build13166; that PID was absent at post-run census, so its exact start is not independently witnessed. Peer untouched; no performance timing claim.

Preserved artifacts: managed `fak-worker-wt-metalgemm-fa700762879c/_scratch/issue-12919-hil/{receipt.json,supervision.json,test.jsonl,test.log,before-check.json,after-check.json}`. Raw JSONL SHA256ffa347447a0a3742e055919bea9810807a395425220b6a42f1044579fc72b24b. Receipt explicitly records competing work; source candidate remains frozen and unlanded.

Next: independent source-only scoping of Q8 graph projection reduction/accumulation against native grouped GEMV, preserving P4 batching and the original strict oracle. No additional run or kernel/algorithm edit before candidate review. Parent #12349 remains open; #12853 and #12914 evidence remains provenance-distinct.
